x-pack/plugins/security_solution/docs/prebuilt_rules_customization_rfc.md

```diff
@@ -0,0 +1,728 @@
+
+**Security Rules (prebuilt rule assets) mapping**
+
+No changes will be needed either for the `security-rule` [mapping](https://github.com/elastic/kibana/blob/main/x-pack/plugins/security_solution/server/lib/detection_engine/prebuilt_rules/logic/rule_assets/prebuilt_rule_assets_type.ts), our prebuilt rule assets. We currently have mappings for the `rule_id` and `version` fields, since we perform aggregations and filtering on those in the rule install and upgrade endpoints. No additional fields needs to be mapped in this phase.
+
+## Plan for carrying out migrations of rule SOs
+
+### Migration strategy: perform migration on all write operations of the rules
+
+In order to migrate our users rule's data to the new schema, we should perform the migration of rules whenever any endpoint operation that writes/updates a rule saved object.
+
+> In Elasticsearch, the type of rule saved objects is called `alert`.
+
+This migration strategy of write/update operations means that a user's data will be migrated incrementally, and that both types of data (non-migrated and migrated saved objects) will coexist for an indeterminate amount of time. Therefore we have to maintain backwards data compatibility for the non-migrated data types.
+
+The migration logic should take place as a first step within the handler logic of all endpoints that carry out write/update endpoint operations, and the endpoint should return the already-migrated rule(s).
+
+### Rule Management endpoints that should include this migration logic are:
+
+All of the following endpoints either fetch the rule before updating it, or send the rule's params as part of the body in the request. Therefore, in order to carry out the migration, we can check if the `params` fields already contains the `prebuilt` and `customized` fields, create them within the `params` if they don't, and then proceed to saving them to Elasticsearch.
+
+- **Update Rule** - `PUT /rules`: same as Patch Rule.
+    - Used **only** when updating/modifying a single rule via the Rule Details page
+- **Patch Rule** - `PATCH /rules`: should migrate patched rule to prevent the endpoint from returning an outdated (non-migrated) rule.
+    - Used for attaching shared exceptions list to rules
+- **Bulk Update Rules** - `PUT /rules/_bulk_update`: same as for Patch Rule
+    - Deprecated and unused by the UI (might still be used by API users)
+- **Bulk Patch Rules** - `PATCH /rules/_bulk_update`: same as for Patch Rule
+    - Deprecated and unused by the UI (might still be used by API users)
+- **Import Rules** - `POST /rules/_import`: 
+    - if the user imports rules with the updated rule schema (`prebuilt` and `customized` fields are present), then save rule as-is. 
```

Comment:
If only `prebuilt` and `customized` fields are present, wouldn't we need to initialize the `immutable` field if it wasn't provided by the user?

Reply:
This is true, thanks. Added a detailed explanation for this case.